Resize `Panel2D` with `WindowResizeEvent`

antrikshmisri opened the pull request.

antrikshmisri committed:
feat: resize panel when window resizes

antrikshmisri committed:
test files

antrikshmisri committed:
feat: added relevant callbacks

antrikshmisri committed:
fix: use partial to provide the obj, iren to window propagate event

antrikshmisri committed:
temporary print statement for window resize callback

antrikshmisri committed:
feat: resize panel according to precalculated ratio

antrikshmisri committed:
fix: cap the panel size to max size

antrikshmisri committed:
fix: removed use of private methods

antrikshmisri committed:
Merge branch 'master' of https://github.com/antrikshmisri/fury into panel-resize

antrikshmisri committed:
feat: force render on resie event

antrikshmisri committed:
fix: pep8 issues

antrikshmisri committed:
feat: resize from bottom right corner

pep8speaks commented:
Hello @antrikshmisri! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:

* In the file [`fury/ui.py`](https://github.com/fury-gl/fury/blob/5d5792e0e14a14d0ccb860468fb165a6cc2eee89/fury/ui.py):

> [Line 167:13](https://github.com/fury-gl/fury/blob/5d5792e0e14a14d0ccb860468fb165a6cc2eee89/fury/ui.py#L167): [E722](https://duckduckgo.com/?q=pep8%20E722) do not use bare 'except'
> [Line 323:5](https://github.com/fury-gl/fury/blob/5d5792e0e14a14d0ccb860468fb165a6cc2eee89/fury/ui.py#L323): [E301](https://duckduckgo.com/?q=pep8%20E301) expected 1 blank line, found 0

antrikshmisri committed:
fix: removed unused libs

antrikshmisri committed:
feat: user hooks

antrikshmisri committed:
feat: added user hook for panel resizing

antrikshmisri committed:
example for panel resizing

antrikshmisri committed:
fix: pep8 issues

antrikshmisri committed:
fix: rebased w.r.t the UI restructuring

antrikshmisri committed:
deleted redundant files

antrikshmisri committed:
fix: removed unused lib

antrikshmisri committed:
fix: pep8 issues

antrikshmisri commented:
Hey @skoudoro, while rebasing this PR w.r.t the UI restructuring I noticed that you cannot import any class from `fury/ui/elements.py` in `fury/ui/containers.py` because `Panel2D` from `containers` is being imported into `elements` which makes them circularly dependent. This is the error:-
```python
  File "D:\pythonProjects\fury-event\fury\fury\ui\containers.py", line 12, in <module>
    from fury.ui.elements import Button2D
  File "D:\pythonProjects\fury-event\fury\fury\ui\elements.py", line 19, in <module>  
    from fury.ui.containers import Panel2D
ImportError: cannot import name 'Panel2D'
```

antrikshmisri committed:
Merge branch 'master' into panel-resize

antrikshmisri committed:
fix: fixed corner resize bug

skoudoro commented on fury/ui/containers.py:
```diff
@@ -5,9 +5,11 @@
 
 from fury.io import load_image
 from fury.ui.core import UI, Rectangle2D, TextBlock2D
+from fury.ui.elements import Button2D
```
now, it is in `ui.core`

skoudoro commented on fury/ui/containers.py:
```diff
@@ -24,7 +26,7 @@ class Panel2D(UI):
 
     def __init__(self, size, position=(0, 0), color=(0.1, 0.1, 0.1),
-                 opacity=0.7, align="left"):
+                 opacity=0.7, align="left", corner_resizable=True):
```
`resizable=True` instead

antrikshmisri replied:
Sure, makes sense

skoudoro commented on fury/ui/containers.py:
```diff
@@ -39,29 +41,45 @@ def __init__(self, size, position=(0, 0), color=(0.1, 0.1, 0.1),
         self.element_offsets = []
         self.background = Rectangle2D()
+        self.resize_button = Button2D(icon_fnames=[('resize_icon',
+                                         'https://i.imgur.com/RQF9wLB.png')])
```
indentation to improve

skoudoro commented on fury/ui/containers.py:
```diff
@@ -204,6 +233,35 @@ def left_button_dragged(self, i_ren, _obj, _panel2d_object):
             self.position = new_position
         i_ren.force_render()
 
+    def window_resize(self, i_ren, _obj, panel2d_object):
```
Please, add docstring

antrikshmisri replied:
Done

skoudoro commented on fury/ui/containers.py:
```diff
@@ -204,6 +233,35 @@ def left_button_dragged(self, i_ren, _obj, _panel2d_object):
+        self.panel_resize_callback(i_ren, panel2d_object)
+        i_ren.force_render()
+
+    def panel_resize_callback(self, i_ren, panel2d_object):
```
please, add docstring

antrikshmisri replied:
Done.

skoudoro commented on fury/ui/core.py:
```diff
@@ -324,6 +324,20 @@ def mouse_move_callback(i_ren, obj, self):
         self.on_key_press(i_ren, obj, self)
+    
+    @staticmethod
+    def window_event_propagate(window, evt, obj, i_ren):
```
please, add docstring

antrikshmisri replied:
Done

skoudoro commented on fury/ui/core.py:
```diff
@@ -324,6 +324,20 @@ def mouse_move_callback(i_ren, obj, self):
+        i_ren.event.abort()
+
+    @staticmethod
+    def window_propagate_callback(i_ren, obj, self):
```
please add docstring

antrikshmisri replied:
Done

skoudoro reviewed (commented):
Hi @antrikshmisri,

What is the status of this PR? Can you rebase it?  Button2D is now in the `ui.core` so it should be ok.

Also, could you add some tests? 

Thank you

antrikshmisri committed:
Merge branch 'master' into panel-resize

antrikshmisri committed:
fix: addressed comments

antrikshmisri commented:
Hey @skoudoro, thanks for the review. I have addressed all the comments. The tests will be added soon. I need to figure out how to record window invoked events first before I add tests.

skoudoro commented:
ok, thank you for letting us know. it is strange that all CI's are failing, we need to figure out why there is this issue after your change

antrikshmisri commented:
> ok, thank you for letting us know. it is strange that all CI's are failing, we need to figure out why there is this issue after your change

Seems like the tests for Tab UI are failing. Will look into it.

antrikshmisri commented:
> Also, could you add some tests?

As almost all the changes are based on events, I think I need to re-record the events for the existing tests for Panel2D, which can be seen [here](https://github.com/fury-gl/fury/blob/d498b6d1cf7b9594f55a80cfcbc047a9c8fd5bd1/fury/ui/tests/test_core.py#L11). But if I do that it will overwrite the existing recording, is there a way to extend the recording instead of overwriting the whole recording?

skoudoro commented:
> But if I do that it will overwrite the existing recording, is there a way to extend the recording instead of overwriting the whole recording?

No problem if you overwrite it. Make sure to watch first the recording. Then, you can replicate it approximately

skoudoro commented on fury/ui/containers.py:
```diff
@@ -39,29 +41,45 @@ def __init__(self, size, position=(0, 0), color=(0.1, 0.1, 0.1),
             Must take values in [0, 1].
         align : [left, right]
             Alignment of the panel with respect to the overall screen.
+        resizable: bool
```
`bool, optional`

skoudoro commented on fury/ui/containers.py:
```diff
@@ -204,6 +233,68 @@ def left_button_dragged(self, i_ren, _obj, _panel2d_object):
+        i_ren: :class: `CustomInteractorStyle`
+            Custom Interactor
+        obj: :class: `UI`
+            UI element assocaited with the event
```
typo: associated

skoudoro commented on fury/ui/core.py:
```diff
@@ -325,6 +327,41 @@ def mouse_move_callback(i_ren, obj, self):
+        evt: str
+            Window invoked event
+        obj: :class: `UI`
+            UI element where event isto be propagated
```
typo isto

skoudoro reviewed (commented):
Hi @antrikshmisri,

See below for some additional comments.

Nibba2018 commented on fury/ui/containers.py:
```diff
@@ -92,9 +120,10 @@ def resize(self, size):
         ----------
         size : (float, float)
             Panel size (width, height) in pixels.
-
         """
         self.background.resize(size)
+        button_coords = (int(size[0] - self.resize_button.size[0]), 0)
+        self.update_element(self.resize_button, button_coords)
 
```
Do we implement resizing of contained components along with this PR or do we do it in a separate PR? @skoudoro @antrikshmisri . Unfortunately, I am not able to recall.

antrikshmisri replied:
I can't recall either but I think it would make sense to do that in a seperate PR.

Nibba2018 commented on fury/ui/containers.py:
```diff
@@ -204,6 +233,68 @@ def left_button_dragged(self, i_ren, _obj, _panel2d_object):
+        panel2d_object: :class: `Panel2D`
+            Instance of the Panel2D
+        """
+        _window_size = i_ren.GetInteractor().GetSize()
+        _new_size = self.size_ratio * _window_size
+
+        self.re_align(_window_size)
+        self.resize(np.clip(_new_size, 0, self.max_size))
+        i_ren.force_render()
```
Good :+1:

Nibba2018 commented on fury/ui/core.py:
```diff
@@ -325,6 +327,41 @@ def mouse_move_callback(i_ren, obj, self):
+        i_ren: :class: `CustomInteractorStyle`
+            Custom Interactor associated with the window
+        """
+        if not hasattr(obj, 'actor'):
+            raise AttributeError(
+                f'{obj} doesnt have an actor associated with it')
```
just `getattr(obj, 'actor')` for this entire block will work.

antrikshmisri replied:
Just provides a more insightful error message.

Nibba2018 reviewed (commented):
Hello @antrikshmisri , Thank You for this PR, I have added some comments below.

antrikshmisri commented on fury/ui/containers.py:
```diff
@@ -39,29 +41,45 @@ def __init__(self, size, position=(0, 0), color=(0.1, 0.1, 0.1),
         self._elements = []
         self.element_offsets = []
         self.background = Rectangle2D()
+        self.resize_button = Button2D(icon_fnames=[('resize_icon',
+                                      'https://i.imgur.com/RQF9wLB.png')])
+
         self.add_element(self.background, (0, 0))
+        self.add_element(self.resize_button, (0, 0))
+
+        if not self.resizable:
+            self.resize_button.set_visibility(False)
 
         # Add default events listener for this UI component.
         self.background.on_left_mouse_button_pressed = self.left_button_pressed
         self.background.on_left_mouse_button_dragged = self.left_button_dragged
+        self.background.on_window_propagate = self.window_resize
+        self.resize_button.on_left_mouse_button_pressed = \
+            self.left_button_pressed
+
+        self.resize_button.on_left_mouse_button_dragged = self.corner_resize
 
```
Sure, this makes sense.

antrikshmisri commented on fury/ui/containers.py:
```diff
@@ -77,8 +95,18 @@ def _add_to_scene(self, scene):
+        window_size = window.GetSize()
+        panel_size = self.size
+        self.size_ratio = panel_size / window_size
+
+        window.AddObserver('WindowResizeEvent',
+                           partial(self.window_event_propagate,
+                                   obj=self.background, i_ren=i_ren))
+
```
`partial` is being used to pass additional arguments (actor associated with the `panel`, `interactor`) to the `window_event_propagate` method

antrikshmisri committed:
fix: addressed comments

antrikshmisri committed:
fix: pep8 issues

antrikshmisri committed:
fix: fixed docstrings

antrikshmisri commented:
Hey @skoudoro @Nibba2018, I am not sure how to test callbacks like [these](https://github.com/fury-gl/fury/pull/446/files#diff-53918b2a3199d9c46ad955f4e1c36bd0c776fd5228674415f4be2306080f23bdR258) or  [these](https://github.com/fury-gl/fury/pull/446/files#diff-53918b2a3199d9c46ad955f4e1c36bd0c776fd5228674415f4be2306080f23bdR239). Also, the events invoked by window are not being recorded even though the same event is being propagated to the respective actor.

antrikshmisri committed:
tests: added tests for resizeable flag, size_ratio

antrikshmisri committed:
updated tests recording

antrikshmisri committed:
fix: changed resizable flag to False by default

antrikshmisri committed:
fix: changed events to WindowResizeEvent

antrikshmisri committed:
fix: re-recorded the tests

antrikshmisri commented:
Hey @skoudoro @Nibba2018, the tests for tab_ui are passing now Not sure what was happening earlier but now everything is fine. Also, the `WindowResizeEvent` is still not being counted even though it has been added to the `EventCounter`. Also, the invokation/calling of the `WindowResizeEvent` is really slow, for 1-2 `Panel2D` its fine but from 4-5+ `Panel2D` it becomes painfully slow.  I think this is because the event is being propagated to many actors at once instead of aborting at panel's actor. What do you guys think?

skoudoro commented:
ok, Thank you for the feedback. Please, keep an example on a side to show how slow it is. 

For now, we won't use  `WindowResizeEvent `  which means this PR is almost ready to be merged

antrikshmisri committed:
updated tests

antrikshmisri committed:
fix: added note about performace issues with window resize

antrikshmisri committed:
fix: fixed conflicts, re-recorded tests

antrikshmisri commented:
Hey @skoudoro @Nibba2018, all the proposed changes have been added. PTAL.

antrikshmisri commented:
Hey @skoudoro @Nibba2018, just a quick note the icon is fetched from a URL, if there is no internet connection this will error out. Possible fixes are to add the resize icon to the FURY icon dataset or add a broken link icon and if there is no internet connection the icon will be changed to that.

ganimtron-10 commented:
Hey @antrikshmisri ,
Can you please rebase this PR?

ganimtron-10 commented:
As we can see below when we resize the panel horizontally it works fine, but when we resize it vertically it seem to reposition itself as sown below. Is this expected?

https://github.com/fury-gl/fury/assets/64432063/6a8d3580-4e28-47b7-b0a3-9d048634bff0

skoudoro commented:
No, it is not expected @ganimtron-10? You might need to look what is going on
